cylc/uiserver/graphql/tornado_ws.py

```diff
@@ -0,0 +1,399 @@
+        pass
+
+    async def on_stop(self, connection_context, op_id):
+        return await connection_context.unsubscribe(op_id)
+
+    async def on_close(self, connection_context):
+        return await connection_context.unsubscribe_all()
+
+    async def handle(self, ws, request_context=None):
+        await asyncio.shield(self._handle(ws, request_context))
+
+    async def _handle(self, ws, request_context=None):
+        connection_context = TornadoConnectionContext(ws, request_context)
+        await self.on_open(connection_context)
+        while True:
+            message = None
+            try:
+                if connection_context.closed:
+                    raise ConnectionClosedException()
+                message = await connection_context.receive()
+            except QueueEmpty:
+                pass
+            except ConnectionClosedException:
+                break
+            if message:
+                await self.on_message(connection_context, message)
+            else:
+                await asyncio.sleep(NO_MSG_DELAY)
+
+        await self.on_close(connection_context)
+
+    async def on_start(self, connection_context, op_id, params):
+        # Attempt to unsubscribe first in case we already have a subscription
+        # with this id.
+        await connection_context.unsubscribe(op_id)
+
+        params['kwargs']['root_value'] = op_id
+        execution_result = await self.execute(params)
+        iterator = None
+        try:
+            if isawaitable(execution_result):
+                execution_result = await execution_result
+            if not hasattr(execution_result, '__aiter__'):
+                await self.send_execution_result(
+                    connection_context, op_id, execution_result)
+            else:
+                iterator = execution_result.__aiter__()
+                connection_context.register_operation(op_id, iterator)
+                async for single_result in iterator:
+                    if not connection_context.has_operation(op_id):
+                        break
+                    await self.send_execution_result(
+                        connection_context, op_id, single_result)
+        except (GeneratorExit, asyncio.CancelledError):
+            raise
+        except Exception as e:
+            await self.send_error(connection_context, op_id, e)
+        finally:
+            if iterator:
+                await iterator.aclose()
+        await self.send_message(connection_context, op_id, GQL_COMPLETE)
+        await connection_context.unsubscribe(op_id)
+        await self.on_operation_complete(connection_context, op_id)
+
+    async def send_message(
+        self, connection_context, op_id=None, op_type=None, payload=None
+    ):
+        if op_id is None or connection_context.has_operation(op_id):
+            message = self.build_message(op_id, op_type, payload)
+            return await connection_context.send(message)
```

Comment:
Not a new problem in this PR, but I have figured out why we don't have any record of subscription errors anywhere. It's because this function doesn't do anything if the operation ID is not registered in the connection context, which it isn't for a graphql subscription error. 

But why doesn't it do anything? As far as I can tell, neither `self.build_message` nor `connection_context.send` make use of the iterator registered with the operation ID, so why not just do?

```suggestion
        message = self.build_message(op_id, op_type, payload)
        return await connection_context.send(message)
```

To test out what I'm talking about, try running the GUI with a typo in the `Workflows.vue` subscription, or by doing a GraphiQL subscription like
```gql
subscription App {
  deltas {
    id
    borked
  }
}
```